com.sap.cloud.lm.sl.cf.process/src/main/java/com/sap/cloud/lm/sl/cf/process/util/JarSignatureVerifier.java

```diff
@@ -0,0 +1,175 @@
+public class JarSignatureVerifier {
+
+    private static final String META_INF = "META-INF";
+    private static final String CN = "CN";
+    private static final int BUFFER_SIZE = 8 * 1024;
+
+    public void verify(List<X509Certificate> targetCertificates, String certificateCN, URL jarURL) {
+        try {
+            JarFile jarFile = openJarFile(jarURL);
+            List<JarEntry> jarEntries = getJarEntries(jarFile);
+            verifyJarEntries(targetCertificates, jarEntries, certificateCN);
+        } catch (IOException | SecurityException | SLException e) {
+            throw new SLException(MessageFormat.format(Messages.COULD_NOT_VERIFY_ARCHIVE_SIGNATURE, e.getMessage()), e);
+        }
+    }
+
+    private JarFile openJarFile(URL jarFileUrl) throws IOException {
+        JarURLConnection connection = (JarURLConnection) toJarUrl(jarFileUrl).openConnection();
+        return connection.getJarFile();
+    }
+
+    private URL toJarUrl(URL jarUrl) throws MalformedURLException {
+        return new URL("jar:" + jarUrl.toString() + "!/");
+    }
+
+    private List<JarEntry> getJarEntries(JarFile jarFile) throws IOException {
+        List<JarEntry> jarEntries = new ArrayList<>();
+        Enumeration<JarEntry> entries = jarFile.entries();
+        while (entries.hasMoreElements()) {
+            JarEntry jarEntry = readJarEntry(jarFile, entries.nextElement());
+            if (!jarEntry.isDirectory()) {
+                jarEntries.add(jarEntry);
+            }
+        }
+        return jarEntries;
+    }
+
+    private JarEntry readJarEntry(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        FileUtils.validatePath(jarEntry.getName());
+        verifySignature(jarFile, jarEntry);
+        return jarEntry;
+    }
+
+    private void verifySignature(JarFile jarFile, JarEntry jarEntry) throws IOException {
+        byte[] buffer = new byte[BUFFER_SIZE];
+        try (InputStream inputStream = jarFile.getInputStream(jarEntry)) {
+            while (inputStream.read(buffer, 0, buffer.length) != -1) {
+                // Just read. This will throw a SecurityException if a signature/digest check fails.
+            }
+        }
+    }
+
+    private void verifyJarEntries(List<X509Certificate> targetCertificates, List<JarEntry> jarEntries, String certificateCN) {
+        List<String> filesWithoutCertificates = new ArrayList<>();
+        for (JarEntry jarEntry : jarEntries) {
+            Certificate[] entryCertificates = jarEntry.getCertificates();
+            if (entryCertificates == null || entryCertificates.length == 0) {
+                verifyIsMetaInformation(jarEntry, filesWithoutCertificates);
+                continue;
+            }
+            verifyJarEntryIsSigned(targetCertificates, entryCertificates, certificateCN);
+        }
+        if (!filesWithoutCertificates.isEmpty()) {
+            throw new SLException(MessageFormat.format(Messages.ARCHIVE_CONTAINS_UNSIGNED_FILES, filesWithoutCertificates));
```

Comment:
Print the list of files only if the archive is signed partially. Use something of the sort:
```
List<JarEntry> nonMetaInformationJarEntries = jarEntries.stream()
                                                        .filter(jarEntry -> !isMetaInformation(jarEntry))
                                                        .collect(Collectors.toList());

List<JarEntry> nonSignedJarEntries = jarEntries.stream()
                                               .filter(jarEntry -> { // Extract the body of this filter in a separate method.
                                                   Certificate[] certificates = jarEntry.getCertificates();
                                                   return certificates == null || certificates.length == 0;
                                                })
                                               .collect(Collectors.toList());

if (!nonSignedJarEntries.isEmpty()) {
    if (nonMetaInformationJarEntries.size() == nonSignedJarEntries.size() {
        throw new SLException("The archive is not signed");
    }
    throw new SLException("The archive contains unsigned files: {0}", String.join("\n", nonSignedJarEntries));
}
```